packages/block-library/src/heading/index.js

```diff
@@ -86,4 +90,21 @@ if ( window.__experimentalContentOnlyInspectorFields ) {
 	};
 }
 
-export const init = () => initBlock( { name, metadata, settings } );
+export const init = () => {
+	const block = initBlock( { name, metadata, settings } );
+
+	// Unregister heading level variations based on `levelOptions` attribute.
+	// This is for backwards compatibility, as extenders can now unregister the
+	// variation directly: `wp.blocks.unregisterBlockVariation( 'core/heading', 'h1' )`.
+	const levelOptions =
+		getBlockType( name )?.attributes?.levelOptions?.default;
+	if ( levelOptions ) {
+		[ 1, 2, 3, 4, 5, 6 ].forEach( ( level ) => {
+			if ( ! levelOptions.includes( level ) ) {
+				unregisterBlockVariation( name, `h${ level }` );
+			}
+		} );
+	}
```

Comment:
This works and I think it's totally fine for now but: 

changing a "default" attribute value to define what levels are available seems like a bad thing to start with. I wonder if we should deprecate that attribute. Unregistering a variation seems like a way better approach.

Reply:
That would break sites that rely on the default options attribute to remove the heading levels.

I think @fabiankaegy is also suggesting keeping the toolbar dropdown item - https://github.com/WordPress/gutenberg/pull/73823#issuecomment-3628129696.